### ISSUE TITLE
Build failed under Mac OS X Mountain Lion

### ISSUE THREAD

**Issue by peterjc** [brentp/pybloomfaster#1]
Fails to build for me, errors from clang making cybloomfaster.so as shown below, using latest commit d5dc05ee082fe911f4c3e7231f47dde6c1219375 from github:

<pre>
$ git clone https://github.com/brentp/pybloomfaster.git
Cloning into pybloomfaster...
remote: Counting objects: 88, done.
remote: Compressing objects: 100% (76/76), done.
remote: Total 88 (delta 35), reused 9 (delta 4)
Unpacking objects: 100% (88/88), done.$
peter-cocks-mac-pro:repositories pjcock$ cd pybloomfaster/
$ ls
README.rst  bloomfaster examples    setup.py    src
$ python setup.py build
running build
running build_py
creating build
creating build/lib.macosx-10.8-intel-2.7
creating build/lib.macosx-10.8-intel-2.7/bloomfaster
copying bloomfaster/__init__.py -> build/lib.macosx-10.8-intel-2.7/bloomfaster
running build_ext
building 'cybloomfaster' extension
creating build/temp.macosx-10.8-intel-2.7
creating build/temp.macosx-10.8-intel-2.7/src
clang -fno-strict-aliasing -fno-common -dynamic -g -Os -pipe -fno-common -fno-strict-aliasing -fwrapv -mno-fused-madd -DENABLE_DTRACE -DMACOSX -DNDEBUG -Wall -Wstrict-prototypes -Wshorten-64-to-32 -DNDEBUG -g -Os -Wall -Wstrict-prototypes -DENABLE_DTRACE -arch i386 -arch x86_64 -pipe -Isrc/ -I/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -c src/cybloomfaster.c -o build/temp.macosx-10.8-intel-2.7/src/cybloomfaster.o
clang: warning: argument unused during compilation: '-mno-fused-madd'
src/cybloomfaster.c:567:286: warning: incompatible pointer to integer conversion passing 'PyObject *' (aka 'struct _object *') to parameter of type 'int';
  ...__pyx_v_n_elements, __pyx_v_error_rate, ((struct __pyx_obj_13cybloomfaster_Elf *)__pyx_v_self)->_bloom->stat.ideal_hashes, NULL, __pyx_int_1);
                                                                                                                                      ^~~~~~~~~~~
src/bloom.h:91:36: note: passing argument to parameter 'flags' here
               int hashes,hash_t hash,int flags);
                                          ^
1 warning generated.
src/cybloomfaster.c:567:286: warning: incompatible pointer to integer conversion passing 'PyObject *' (aka 'struct _object *') to parameter of type 'int';
  ...__pyx_v_n_elements, __pyx_v_error_rate, ((struct __pyx_obj_13cybloomfaster_Elf *)__pyx_v_self)->_bloom->stat.ideal_hashes, NULL, __pyx_int_1);
                                                                                                                                      ^~~~~~~~~~~
src/bloom.h:91:36: note: passing argument to parameter 'flags' here
               int hashes,hash_t hash,int flags);
                                          ^
1 warning generated.
clang -fno-strict-aliasing -fno-common -dynamic -g -Os -pipe -fno-common -fno-strict-aliasing -fwrapv -mno-fused-madd -DENABLE_DTRACE -DMACOSX -DNDEBUG -Wall -Wstrict-prototypes -Wshorten-64-to-32 -DNDEBUG -g -Os -Wall -Wstrict-prototypes -DENABLE_DTRACE -arch i386 -arch x86_64 -pipe -Isrc/ -I/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -c src/bloom.c -o build/temp.macosx-10.8-intel-2.7/src/bloom.o
clang: warning: argument unused during compilation: '-mno-fused-madd'
src/bloom.c:125:20: warning: implicit conversion loses integer precision: 'unsigned long long' to 'int' [-Wshorten-64-to-32]
        ret = (BIGNUM)abs(bloom->hash(newstr)) % (BIGNUM)bloom->stat.elements;
                      ~~~ ^~~~~~~~~~~~~~~~~~~
src/bloom.c:192:44: warning: implicit conversion loses integer precision: 'unsigned long long' to 'size_t' (aka 'unsigned long') [-Wshorten-64-to-32]
        char *ret = (char *)malloc(sizeof(char *) * (bm->stat.elements + 2));
                            ~~~~~~ ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
src/bloom.c:125:20: warning: implicit conversion loses integer precision: 'unsigned long long' to 'int' [-Wshorten-64-to-32]
        ret = (BIGNUM)abs(bloom->hash(newstr)) % (BIGNUM)bloom->stat.elements;
                      ~~~ ^~~~~~~~~~~~~~~~~~~
src/bloom.c:226:11: warning: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'int' [-Wshorten-64-to-32]
        if ((r = fwrite(b->vector,sizeof(char)*elems,1,fp)) != 1) {
               ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
clang -fno-strict-aliasing -fno-common -dynamic -g -Os -pipe -fno-common -fno-strict-aliasing -fwrapv -mno-fused-madd -DENABLE_DTRACE -DMACOSX -DNDEBUG -Wall -Wstrict-prototypes -Wshorten-64-to-32 -DNDEBUG -g -Os -Wall -Wstrict-prototypes -DENABLE_DTRACE -arch i386 -arch x86_64 -pipe -Isrc/ -I/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -c src/suggestions.c -o build/temp.macosx-10.8-intel-2.7/src/suggestions.o
clang: warning: argument unused during compilation: '-mno-fused-madd'
src/suggestions.c:37:40: warning: implicit conversion loses integer precision: 'unsigned long long' to 'int' [-Wshorten-64-to-32]
  stat->ideal_hashes = 9*stat->elements/(13*stat->capacity);
                     ~ ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
1 warning generated.
src/suggestions.c:37:40: warning: implicit conversion loses integer precision: 'unsigned long long' to 'int' [-Wshorten-64-to-32]
  stat->ideal_hashes = 9*stat->elements/(13*stat->capacity);
                     ~ ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
1 warning generated.
clang -bundle -undefined dynamic_lookup -Wl,-F. -arch i386 -arch x86_64 build/temp.macosx-10.8-intel-2.7/src/cybloomfaster.o build/temp.macosx-10.8-intel-2.7/src/bloom.o build/temp.macosx-10.8-intel-2.7/src/suggestions.o src/bloom.h -o build/lib.macosx-10.8-intel-2.7/cybloomfaster.so
clang: error: cannot use 'precompiled-header' output with multiple -arch options
clang: error: cannot specify -o when generating multiple output files
error: command 'clang' failed with exit status 1
$ which python
/usr/bin/python
$ python
Python 2.7.2 (default, Jun 20 2012, 16:23:33) 
[GCC 4.2.1 Compatible Apple Clang 4.0 (tags/Apple/clang-418.0.60)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> quit()
</pre>


My guesses about removing the -arch flags didn't solve it:

<pre>
$ clang -bundle -undefined dynamic_lookup -Wl,-F. -arch i386 -arch x86_64 build/temp.macosx-10.8-intel-2.7/src/cybloomfaster.o build/temp.macosx-10.8-intel-2.7/src/bloom.o build/temp.macosx-10.8-intel-2.7/src/suggestions.o src/bloom.h -o build/lib.macosx-10.8-intel-2.7/cybloomfaster.so
clang: error: cannot use 'precompiled-header' output with multiple -arch options
clang: error: cannot specify -o when generating multiple output files
$ clang -bundle -undefined dynamic_lookup -Wl,-F. -arch x86_64 build/temp.macosx-10.8-intel-2.7/src/cybloomfaster.o build/temp.macosx-10.8-intel-2.7/src/bloom.o build/temp.macosx-10.8-intel-2.7/src/suggestions.o src/bloom.h -o build/lib.macosx-10.8-intel-2.7/cybloomfaster.so
clang: error: cannot specify -o when generating multiple output files
$ clang -bundle -undefined dynamic_lookup -Wl,-F. build/temp.macosx-10.8-intel-2.7/src/cybloomfaster.o build/temp.macosx-10.8-intel-2.7/src/bloom.o build/temp.macosx-10.8-intel-2.7/src/suggestions.o src/bloom.h -o build/lib.macosx-10.8-intel-2.7/cybloomfaster.so
clang: error: cannot specify -o when generating multiple output files
</pre>

**Comment by peterjc** [comment]
I had to re-run cython to regenerate cybloomfaster.c to see any effect, but that commit has fixed several warnings at least:

<pre>
$ python setup.py build
running build
running build_py
running build_ext
building 'cybloomfaster' extension
creating build/temp.macosx-10.8-intel-2.7
creating build/temp.macosx-10.8-intel-2.7/src
clang -fno-strict-aliasing -fno-common -dynamic -g -Os -pipe -fno-common -fno-strict-aliasing -fwrapv -mno-fused-madd -DENABLE_DTRACE -DMACOSX -DNDEBUG -Wall -Wstrict-prototypes -Wshorten-64-to-32 -DNDEBUG -g -Os -Wall -Wstrict-prototypes -DENABLE_DTRACE -arch i386 -arch x86_64 -pipe -Isrc/ -I/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -c src/cybloomfaster.c -o build/temp.macosx-10.8-intel-2.7/src/cybloomfaster.o
clang: warning: argument unused during compilation: '-mno-fused-madd'
clang -fno-strict-aliasing -fno-common -dynamic -g -Os -pipe -fno-common -fno-strict-aliasing -fwrapv -mno-fused-madd -DENABLE_DTRACE -DMACOSX -DNDEBUG -Wall -Wstrict-prototypes -Wshorten-64-to-32 -DNDEBUG -g -Os -Wall -Wstrict-prototypes -DENABLE_DTRACE -arch i386 -arch x86_64 -pipe -Isrc/ -I/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -c src/bloom.c -o build/temp.macosx-10.8-intel-2.7/src/bloom.o
clang: warning: argument unused during compilation: '-mno-fused-madd'
src/bloom.c:125:20: warning: implicit conversion loses integer precision: 'unsigned long long' to 'int' [-Wshorten-64-to-32]
        ret = (BIGNUM)abs(bloom->hash(newstr)) % (BIGNUM)bloom->stat.elements;
                      ~~~ ^~~~~~~~~~~~~~~~~~~
src/bloom.c:192:44: warning: implicit conversion loses integer precision: 'unsigned long long' to 'size_t' (aka 'unsigned long') [-Wshorten-64-to-32]
        char *ret = (char *)malloc(sizeof(char *) * (bm->stat.elements + 2));
                            ~~~~~~ ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
src/bloom.c:125:20: warning: implicit conversion loses integer precision: 'unsigned long long' to 'int' [-Wshorten-64-to-32]
        ret = (BIGNUM)abs(bloom->hash(newstr)) % (BIGNUM)bloom->stat.elements;
                      ~~~ ^~~~~~~~~~~~~~~~~~~
src/bloom.c:226:11: warning: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'int' [-Wshorten-64-to-32]
        if ((r = fwrite(b->vector,sizeof(char)*elems,1,fp)) != 1) {
               ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
clang -fno-strict-aliasing -fno-common -dynamic -g -Os -pipe -fno-common -fno-strict-aliasing -fwrapv -mno-fused-madd -DENABLE_DTRACE -DMACOSX -DNDEBUG -Wall -Wstrict-prototypes -Wshorten-64-to-32 -DNDEBUG -g -Os -Wall -Wstrict-prototypes -DENABLE_DTRACE -arch i386 -arch x86_64 -pipe -Isrc/ -I/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -c src/suggestions.c -o build/temp.macosx-10.8-intel-2.7/src/suggestions.o
clang: warning: argument unused during compilation: '-mno-fused-madd'
src/suggestions.c:37:40: warning: implicit conversion loses integer precision: 'unsigned long long' to 'int' [-Wshorten-64-to-32]
  stat->ideal_hashes = 9*stat->elements/(13*stat->capacity);
                     ~ ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
1 warning generated.
src/suggestions.c:37:40: warning: implicit conversion loses integer precision: 'unsigned long long' to 'int' [-Wshorten-64-to-32]
  stat->ideal_hashes = 9*stat->elements/(13*stat->capacity);
                     ~ ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
1 warning generated.
clang -bundle -undefined dynamic_lookup -Wl,-F. -arch i386 -arch x86_64 build/temp.macosx-10.8-intel-2.7/src/cybloomfaster.o build/temp.macosx-10.8-intel-2.7/src/bloom.o build/temp.macosx-10.8-intel-2.7/src/suggestions.o src/bloom.h -o build/lib.macosx-10.8-intel-2.7/cybloomfaster.so
clang: error: cannot use 'precompiled-header' output with multiple -arch options
clang: error: cannot specify -o when generating multiple output files
error: command 'clang' failed with exit status 1
</pre>

This is probably some clang/Apple subtlety introduced in Mac OS X 10.8 Mountain Lion :(

For clarity, the failing line once more:

<pre>$ clang -bundle -undefined dynamic_lookup -Wl,-F. -arch i386 -arch x86_64 build/temp.macosx-10.8-intel-2.7/src/cybloomfaster.o build/temp.macosx-10.8-intel-2.7/src/bloom.o build/temp.macosx-10.8-intel-2.7/src/suggestions.o src/bloom.h -o build/lib.macosx-10.8-intel-2.7/cybloomfaster.so
clang: error: cannot use 'precompiled-header' output with multiple -arch options
clang: error: cannot specify -o when generating multiple output files
</pre>

As noted above, removing -arch i386 (since this is a 64bit machine) or both the -arch settings does resolve this error:

<pre>clang: error: cannot use 'precompiled-header' output with multiple -arch options</pre>

However, I don't understand what this error is trying to say:

<pre>clang: error: cannot specify -o when generating multiple output files</pre>

**Comment by brentp** [comment]
thanks for trying again. I'm not sure what to tell you about that.

**Comment by peterjc** [comment]
I see you committed the updated C file: bf7e0072b976463c714924a9ef23ff6e4a00eb13

I had earlier used Cython version 0.17, so I retested with your C file from Cython 0.16, but saw no change to this build error.

I guess we'll have to leave the issue open for now then... thanks though.

**Comment by brentp** [comment]
closed in pull #2 by @mdshw5
